python_modules/dagster/dagster/_core/definitions/cacheable_assets.py

```diff
@@ -0,0 +1,106 @@
+from typing import AbstractSet, Any, Mapping, NamedTuple, Optional, Sequence
+
+import dagster._check as check
+import dagster._seven as seven
+from dagster._core.definitions.assets import AssetsDefinition
+from dagster._core.definitions.events import AssetKey, CoercibleToAssetKeyPrefix
+from dagster._core.definitions.metadata import MetadataUserInput
+from dagster._serdes import whitelist_for_serdes
+from dagster._utils import frozendict, frozenlist, make_readonly_value
+
+
+@whitelist_for_serdes
+class CachedAssetsData(
+    NamedTuple(
+        "_AssetsDefinitionMetadata",
+        [
+            ("keys_by_input_name", Optional[Mapping[str, AssetKey]]),
+            ("keys_by_output_name", Optional[Mapping[str, AssetKey]]),
+            ("internal_asset_deps", Optional[Mapping[str, AbstractSet[AssetKey]]]),
+            ("group_name", Optional[str]),
+            ("metadata_by_output_name", Optional[Mapping[str, MetadataUserInput]]),
+            ("key_prefix", Optional[CoercibleToAssetKeyPrefix]),
+            ("can_subset", bool),
+            ("extra_metadata", Optional[Mapping[Any, Any]]),
+        ],
+    )
+):
```

Comment:
do we expect users to produce these or just us / library authors? Are you confident this data structure is right for this? Changing this can be a pain since we are persisting it. Having it be in the execution plan snapshots mean we will be loading this anytime some one looks at the associated run in dagit.

Reply:
I think there will be some advanced users that will want to use this behavior (essentially for home-grown libraries), but for the most part we'll be the ones creating these things

I'm confident that this is a reasonably accurate structure (it mirrors the arguments of the non-experimental AssetsDefinition.from_op), but I'm definitely not 100% confident in it.